sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/ScanTaskReader.java

```diff
@@ -183,8 +187,12 @@ public boolean advance() throws IOException {
       }
       GenericDeleteFilter deleteFilter =
           new GenericDeleteFilter(checkStateNotNull(io), fileTask, fileTask.schema(), project);
-      currentIterator = deleteFilter.filter(iterable).iterator();
-
+      iterable = deleteFilter.filter(iterable);
+      if (filter != null && filter.op() != Expression.Operation.TRUE) {
```

Comment:
Can we cache the evaluator instead of creating per element/task ? (also for the other location)

Reply:
Done